File: 프로그래머스/3/60063. 블록 이동하기/블록 이동하기.js
const getKey = (x, y, x1, y1) => {
    const sort = [[x,y],[x1,y1]].sort((a,b) => {
        if(a[0] !== b[0]) {
            return a[0] - b[0];
        }
        return a[1] - b[1];
    });

    return `${sort[0][0]},${sort[0][1]},${sort[1][0]},${sort[1][1]}`;
}

const dx = [1,0,-1, 0];
const dy = [0,1,0,-1];
const check = (p1, p2, board) => {
    const result = [];
    const [x1, y1] = p1;
    const [x2, y2] = p2;

    for(let i = 0; i <dx.length; i++) {
        const nx1 = x1 + dx[i];
        const ny1 = y1 + dy[i];
        const nx2 = x2 + dx[i];
        const ny2 = y2 + dy[i];
        if(nx1 >= 0 && nx1 < board.length && nx2 >= 0 && nx2 < board.length){
            if(ny1 >= 0 && ny1 < board.length && ny2 >= 0 && ny2 < board.length) {
                if(board[nx1][ny1] === 0 && board[nx2][ny2] === 0){
                    result.push([[nx1,ny1],[nx2,ny2]]);

                    if(x1 === x2 && dx[i] !== 0) { // 가로
                        result.push([[x1,y1],[nx1,ny1]]);
                        result.push([[x2,y2],[nx2,ny2]]);
                    } else if(y1 === y2 && dy[i] !== 0) { // 세로
                        result.push([[x1,y1],[nx1,ny1]]);
                        result.push([[x2,y2],[nx2,ny2]]);
                    }
                }   
            }
        }
    }

    return result;
}
const isLastNode = (p1, p2, board) => {
    const x = board.length -1;
    const y = board[0].length -1;

    if(p1[0] === x && p1[1] === y) {
        return true;
    }
    if(p2[0] === x && p2[1] === y) {
        return true;
    }
    return false;
}
// visited의 판별 기준 map
function solution(board) {
    var answer = Number.POSITIVE_INFINITY;

    const vmap = new Map();
    vmap.set(getKey(0,0,0,1),0);

    const dot1 = [0,0];
    const dot2 = [0,1];
    // bfs
    const q = [[dot1, dot2]];

    while(q.length) {
        const [p1, p2] = q.shift();
        const prevKey = getKey(p1[0], p1[1],p2[0], p2[1]);
        const prevCnt = vmap.get(prevKey);


        if(isLastNode(p1,p2, board)) {
            return prevCnt;
        }

        // 8가지에 대해서 가능한 것 확인 후 삽입.
        const result = check(p1,p2,board);
        result.forEach((point) => {
            const [point1 , point2] = point;
            const key = getKey(point1[0], point1[1],point2[0], point2[1]);
            const cnt = vmap.get(key);
            if(cnt === undefined || prevCnt + 1 < cnt) {
                vmap.set(key, prevCnt + 1);
                q.push([point1, point2]);
            }
        });
    }


}

File: 프로그래머스/3/60063. 블록 이동하기/블록 이동하기.java
import java.util.*;

class Node{
    int y1;
    int x1;
    int y2;
    int x2;
    int time;
    
    public Node(int y1, int x1, int y2, int x2, int time){
        this.y1 = y1;
        this.x1 = x1;
        this.y2 = y2;
        this.x2 = x2;
        this.time = time;
    }
}

class Solution {
    
    int[] ud = {1,0,-1,0};
    int[] lr = {0,1,0,-1};
        
    public String getPoint(int y1, int x1, int y2, int x2){
		int[][] p = {{y1, x1}, {y2, x2}};
		Arrays.sort(p, (a,b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            return a[0] - b[0];
        });
		
		return p[0][0] + "," + p[0][1] + "," + p[1][0] + "," + p[1][1];
	}
    
    public int solution(int[][] board) {
        int answer = 0;
        
        Map<String, Integer> map = new HashMap<>();
        int n = board.length;
        int m = board[0].length;
        int INF = Integer.MAX_VALUE;
        answer = INF;
        
        PriorityQueue<Node> q = new PriorityQueue<>(new Comparator<Node>(){
            @Override
            public int compare(Node a, Node b){
                return a.time - b.time;
            }
        });
        q.add(new Node(0,1,0,0,0));
        map.put(getPoint(0,1,0,0), 0);
        
        while(!q.isEmpty()){
            Node cur = q.poll();
            
            if(cur.y1 == n-1 && cur.x1 == m-1){
                return cur.time;
            }
            
            if(cur.y2 == n-1 && cur.x2 == m-1){
                return cur.time;
            }
            
            if(cur.time > map.get(getPoint(cur.y1, cur.x1, cur.y2, cur.x2))) continue;
            
            
            for(int i=0; i<4; i++){
                int n_y1 = cur.y1 + ud[i];
                int n_x1 = cur.x1 + lr[i];
                int n_y2 = cur.y2 + ud[i];
                int n_x2 = cur.x2 + lr[i];
                
                if(n_y1 >= n || n_x1 >= m || n_y1 < 0 || n_x1 < 0) continue;
                if(n_y2 >= n || n_x2 >= m || n_y2 < 0 || n_x2 < 0) continue;
                if(board[n_y1][n_x1] == 1 || board[n_y2][n_x2] == 1) continue;
                
                if(cur.time+1 >= map.getOrDefault(getPoint(n_y1, n_x1, n_y2, n_x2), INF)) continue;
                
                q.add(new Node(n_y1, n_x1, n_y2, n_x2, cur.time+1));
                map.put(getPoint(n_y1, n_x1, n_y2, n_x2), cur.time+1);
                
                if(cur.y1 == cur.y2 && ud[i] != 0){
                    if(cur.time+1 < map.getOrDefault(getPoint(cur.y1, cur.x1, n_y1, n_x1), INF)){
                        q.add(new Node(cur.y1, cur.x1, n_y1, n_x1, cur.time+1));
                        map.put(getPoint(cur.y1, cur.x1, n_y1, n_x1), cur.time+1);
                    }
                    
                    if(cur.time+1 < map.getOrDefault(getPoint(cur.y2, cur.x2, n_y2, n_x2), INF)){
                        q.add(new Node(cur.y2, cur.x2, n_y2, n_x2, cur.time+1));
                        map.put(getPoint(cur.y2, cur.x2, n_y2, n_x2), cur.time+1);
                    }
                }
                if(cur.x1 == cur.x2 && lr[i] != 0){
                    if(cur.time+1 < map.getOrDefault(getPoint(cur.y1, cur.x1, n_y1, n_x1), INF)){
                        q.add(new Node(cur.y1, cur.x1, n_y1, n_x1, cur.time+1));
                        map.put(getPoint(cur.y1, cur.x1, n_y1, n_x1), cur.time+1);
                    }
                    
                    if(cur.time+1 < map.getOrDefault(getPoint(cur.y2, cur.x2, n_y2, n_x2), INF)){
                        q.add(new Node(cur.y2, cur.x2, n_y2, n_x2, cur.time+1));
                        map.put(getPoint(cur.y2, cur.x2, n_y2, n_x2), cur.time+1);
                    }
                }
            }
        }
        
        return answer;
    }
}